### PR TITLE
Adding the  missing special methods

### PR DESCRIPTION
**Pull request by soma2000-lang**
Solves https://trello.com/c/QDPhbgM6

### CODE REVIEW

**Review by Aarsh2001 — COMMENTED** [review]
Hey @soma2000-lang, we have updated the task a bit. Kindly have a look at the description, thanks 🙂

**Comment by soma2000-lang** [comment]
Hi @Aarsh2001 Just  had some silly doubt.

1. Could not  find the native shape class for the torch.size.

2. We need to add the special methods to init py file right?

**Review by Aarsh2001 — COMMENTED** [review]
Hey @soma2000-lang, you need to check the methods by printing `dir(torch.Size)` and `dir(tf.TensorShape)`. So it should output something like this(for torch) -:

Screenshot 2023-05-03 at 5 21 22 PM

I guess the way they implement `torch.Size` is a bit different, and it is implemented as a subclass of std::vector<int64_t>, where it overrides several methods to provide additional functionality for working with tensor sizes. You can find the base file [here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/Size.cpp). But it don't think you need to refer it for the implementations. 

What you should do instead is -:
1. check for the intersection of `dunder` methods in `tf.TensorShape` and `torch.Size()`. This is priority number 1
2. implement all other property/class methods as supersets. So these methods would be which don't have a preceding or succeeding underscore. For example `numel` and `count` in the screenshot above. You can open another PR for this. 

Hope this clarifies, if not, feel free to reach out to me 🙂

**Comment by soma2000-lang** [comment]
image

**Comment by Aarsh2001** [comment]
> Hi @Aarsh2001 I have a doubt ,I cannot find reference to some dunder methods like '''**reduce_ex** ''' and ""**format**""" I am also taking reference from [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/tensor_shape.py)
> 
> For the property methods I will open another PR

Hey @soma2000-lang, the `__reduce_ex` method is used for serialising/deserialising objects. I don't think we need to implement this, cause we would be defaulting to the `NativeShape` class anyway. They have handled it in their implementation🙂. Here's an example I ran, which works for `ivy.Shape` -:

```
import ivy
import pickle

ivy.set_backend("tensorflow")
x = ivy.array([2., 3.])
y = ivy.Shape(x.shape)

with open('my_dict.pkl', 'wb') as f:
    # Serialize the dictionary to the file using pickle.dump()
    pickle.dump(y, f)
```
This can be extended  to other backends as well. I don't think we need the `__format__` method either, and a few others like `__init_subclass__`, `__subclasshook__`. I think this PR is a good merge 🙂, requesting a final review from @VedPatwardhan

File: ivy/__init__.py [hunk 1]
```diff
@@ -263,6 +314,29 @@ def __getitem__(self, key):
...
         return len(self._shape) if self._shape is not None else 0
 
+    def __delattr__(self, item):
+        return super().__delattr__(item)
+
+    def __hash__(self):
+        return hash(self._shape)
+
+    def __setattr__(self, item):
+        return super().__setattr__(item)
```
Comment (Aarsh2001) [inline ivy/__init__.py:324]: Also, let's remove this, we don't need it

**Review by Aarsh2001 — CHANGES_REQUESTED** [review]

**Review by VedPatwardhan — APPROVED** [review]
looks good to merge, thanks @soma2000-lang @Aarsh2001 🙂

**Pull request merged**